### PR TITLE
Vortex 5566: Lower active data panel providers

### PR DESCRIPTION
**Pull request by brelsftm**
Created ability to add providers to the lower active data panel.

### CODE REVIEW

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/DetailPanelProvider.java [hunk 1]
```diff
@@ -33,4 +33,4 @@
      * @return a detail panel implementation for the supplied data group.
      */
     DetailPane getDetailPanel(DataGroupInfo dataGroupInfo);
-}
+}
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/DetailPanelProvider.java]: Make sure the file ends a in a newline.

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java [hunk 2]
```diff
@@ -8,23 +8,29 @@
 /**
  * A registry of detail panel providers. Plugins may register a provider with
  * the registry to add additional behavior.
+ * 
+ * @param <T>
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java]: Document what the "T" parameter is for

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java [hunk 3]
```diff
@@ -8,23 +8,29 @@
...
  */
-public interface DetailPanelProviderRegistry
+public interface SimpleRegistry<T>
 {
     /**
      * Registers the supplied provider with the registry.
+     * 
+     * @param <T>
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java]: You shouldn't need this parameter documentation, it's documented at the class level.

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java [hunk 4]
```diff
@@ -8,23 +8,29 @@
...
      * @param pProvider the provider to add to the registry.
      */
-    void registerProvider(DetailPanelProvider pProvider);
+    void registerProvider(T provider);
 
     /**
      * Gets the {@link Collection} of registered {@link DetailPanelProvider}
      * implementations from the registry.
+     * 
+     * @param <T>
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java]: Same as above

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java [hunk 5]
```diff
@@ -108,6 +109,15 @@
...
     private final transient EventListener<DataTypeInfoLoadsToChangeEvent> myLoadsToSubscriber = this::handleLoadsToChange;
 
+    /** The toolbox. */
+    private final Toolbox myToolbox;
+
+    /** The control panel toolbox. **/
+    private final ControlPanelToolbox myCpToolbox;
+
+    /** The provider panel. **/
+    private GridBagPanel myProviderPanel;
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java]: You should be able to make this final

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/ControlPanelToolbox.java [hunk 6]
```diff
@@ -1,42 +1,62 @@
...
+     *
+     * @return the registry in which detail panel providers are registered.
+     */
+    SimpleRegistry<Function<DataTypeInfo, Component>> getLayerControlProviderRegistry();
+
     /**
      * Gets the registry for {@link DetailPanelProvider} implementations. This
      * allows for plugins to register additional providers without injecting
      * dependencies into the core project.
+     *
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/ControlPanelToolbox.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/ControlPanelToolbox.java [hunk 7]
```diff
@@ -1,42 +1,62 @@
...
  * Defines a toolbox for use with control panels.
  */
 public interface ControlPanelToolbox extends PluginToolbox
 {
+
+    /**
+     * Gets the registry for {@link LayerControlProvider} implementations. This
+     * allows for plugins to register additional providers without injecting
+     * dependencies into the core project.
+     *
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/ControlPanelToolbox.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java [hunk 8]
```diff
@@ -33,4 +39,5 @@
      * @param pProperties the properties with which to initialize the registry.
      */
     void initialize(Toolbox pToolbox, Properties pProperties);
+
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/SimpleRegistry.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java [hunk 9]
```diff
@@ -278,11 +299,6 @@ public void setSelected(LayerSelectedEvent selectEvent)
         }
     }
 
-    /**
-     * Builds the color/opacity panel.
-     *
-     * @return the color/opacity panel
-     */
     private GridBagPanel buildColorOpacityPanel()
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java]: Make sure all your private methods are documented with JavaDocs.

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java [hunk 10]
```diff
@@ -822,6 +854,28 @@ private void rebuildMiniStyleBox(DataGroupInfo dgi, DataTypeInfo dti)
...
+    /**
+     * Rebuild the provider panel.
+     * 
+     * @param uniqueProvider the provider component.
+     */
+    private void rebuildProviderLayerControl(Component uniqueProvider)
+    {
+        getProviderPanel().removeAll();
+        GridBagPanel providerPanel = new GridBagPanel();
+        providerPanel.setInsets(0, 0, 0, 0);
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java]: The default is 0,0,0,0, so you shouldn't need to set it to that

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/ControlPanelToolboxImpl.java [hunk 11]
```diff
@@ -86,38 +72,37 @@ protected DetailPanelProviderRegistry createDetailPanelProviderRegistry(Toolbox
     public String getDescription()
     {
         return "A toolbox extension for control panels";
+
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/ControlPanelToolboxImpl.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/ControlPanelToolboxImpl.java [hunk 12]
```diff
@@ -86,38 +72,37 @@ protected DetailPanelProviderRegistry createDetailPanelProviderRegistry(Toolbox
...
-    /**
-     * {@inheritDoc}
-     *
-     * @see io.opensphere.controlpanels.ControlPanelToolbox#setUISpan(ObservableValue) ()
-     */
     @Override
     public void setUISpan(ObservableValue<TimeSpan> uiSpan)
     {
         this.myUISpan = uiSpan;
+
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/ControlPanelToolboxImpl.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/DefaultDetailPanelProvider.java [hunk 13]
```diff
@@ -16,6 +16,7 @@
 /**
  * The default detail panel provider, used to provide detail panels for all data
  * types.
+ * @param <T>
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/DefaultDetailPanelProvider.java]: the @param <T> isn't needed here, as the generic type isn't declared in this class.

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+import io.opensphere.controlpanels.SimpleRegistry;
+import io.opensphere.core.Toolbox;
+import io.opensphere.core.util.collections.New;
+
+/**
+ * A default implementation of the {@link DetailPanelProviderRegistry}, in which
+ * providers are cached.
+ * 
+ * @param <T>
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java]: Document what T is, and what it's used for

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * The collection of registered providers.
+     */
+    private Collection<T> myProviders;
+
+    @Override
+    public void registerProvider(T provider)
+    {
+        myProviders.add(provider);
+
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Override
+    public Collection<T> getProviders()
+    {
+        // TODO Auto-generated method stub
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java]: Remove the TODO

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        return myProviders;
+    }
+
+    @Override
+    public void initialize(Toolbox pToolbox, Properties pProperties)
+    {
+        myProviders = New.collection();
+
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
```
Comment (rkausch) [inline open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/toolbox/SimpleRegistryImpl.java]: Remove the extra blank line (you can get Eclipse to remove extra lines for you on save)

File: open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java (hunk 5 above)
Comment (brelsftm) [reply open-sphere-base/control-panels/src/main/java/io/opensphere/controlpanels/layers/activedata/ActiveLayerControlPanel.java]: I refactored the code so its final now

**Pull request merged**